src/app/frontend/common/services/global/kubeconfig.ts

```diff
@@ -0,0 +1,246 @@
+// Copyright 2017 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {Inject, Injectable, NgZone} from '@angular/core';
+import {HttpErrorResponse} from '@angular/common/http';
+import {Router} from '@angular/router';
+import {CookieService} from 'ngx-cookie-service';
+import {SessionStorageService} from 'ngx-webstorage';
+import {load as loadYaml, dump as dumpYaml} from 'js-yaml';
+
+import {AsKdError, K8SError} from '../../errors/errors';
+import {AuthService} from './authentication';
+import {PluginsConfigService} from './plugin';
+import {LoginSpec} from '@api/backendapi';
+import {KdError} from '@api/frontendapi';
+import {Config, CONFIG_DI_TOKEN} from '../../../index.config';
+
+type Cluster = {
+  name: string;
+  cluster: {
+    server: string;
+    'certificate-authority-data': string;
+    'sidecar-host': string;
+  };
+};
+type User = {
+  name: string;
+  user: {
+    token: string;
+    'client-certificate-data': string;
+    'client-key-data': string;
+  };
+};
+type Context = {
+  name: string;
+  context: {
+    cluster: string;
+    user: string;
+  };
+};
+type Kubeconfig = {
+  apiVersion: string;
+  clusters: Cluster[];
+  contexts: Context[];
+  'current-context': string;
+  users: User[];
+  kind: string;
+  preferences: {};
+};
+
+@Injectable()
+export class KubeconfigService {
+  errors: KdError[] = [];
+
+  constructor(
+    private readonly authService_: AuthService,
+    private readonly pluginConfigService_: PluginsConfigService,
+    private readonly ngZone_: NgZone,
+    private readonly cookies_: CookieService,
+    private readonly sessionStorage_: SessionStorageService,
+    private readonly state_: Router,
+    @Inject(CONFIG_DI_TOKEN) private readonly CONFIG: Config,
+  ) {}
+
+  setKubeconfig(configYaml: string): void {
+    if (!configYaml.length) {
+      return null;
+    }
+    // Store kubeconfig from session storage
+    this.sessionStorage_.store(this.CONFIG.kubeconfigSessionStorageName, configYaml);
```

Comment:
Could you explain why you have picked session storage to keep the data?

Reply:
We have to be very careful with storing the kubeconfig. First, we have to make sure that we can support used kubeconfig file as a login option, not to store invalid kubeconfig files that might contain sensitive certificate information. Also, session storage might not be secure enough to store such sensitive data.

Reply:
Actually, I didn't want to store such a sensitive data as kubeconfig. But to enable to reload page, i.e. to be able to reload kubeconfig by angular service , we need to store somewhere. Also, kubeconfig is not so small to store in cookie. So I think session storage is better place to store kubeconfig while the session than the other.